### PR TITLE
Hide 'Add snap account' button behind an experimental setting

### PR DESCRIPTION
**Pull request by danroc**
## **Description**

In preparation for the stable release of the Snap Accounts API, we would like to hide the "Add Snap account" button from the accounts list, placing it behind an experimental setting.

To achieve this, this PR adds a new section in the experimental settings with a toggle to enable or disable the "Add Snap account" button. It defaults to disabled.

We also track a new event called `Add Snap Account Enabled`.

## **Manual testing steps**

1. Install Flask and onboard your SRP.
2. Go to your accounts list.
3. The "Add Snap account" must not be present.
4. Go to "Settings" > "Experimental".
5. Enable the "Add Snap account" button.
6. Go to your accounts list.
7. The "Add Snap account" must be present.
8. Click on the "Add Snap account"
9. You should see the list of allowed Account Snaps.

### New experimental settings:

image

### Accounts list when "Add Snap account" is disabled:

image

### Accounts list when "Add Snap account" is enabled:

image

## **Related issues**

Fixes: MetaMask/accounts-planning#45

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by gantunesr — COMMENTED** [review]
Just a small comment. It'd be good to also add E2E

**Comment by HowardBraham** [comment]
I tested it with the build downloaded from the PR, and the checkbox didn't move when clicked.  Gave an error about "messages," but not the line number, because it was the full build version.  Tried it again with `yarn start:flask` and it worked fine.  Then tried it with the full build version again and it worked again.  I'm not sure if I can repeat this error, but something might be wrong somewhere, especially if you migrate from an older version.

File: app/_locales/en/messages.json [hunk 1]
```diff
@@ -268,6 +268,12 @@
   "addSnapAccountModalDescription": {
     "message": "Discover options to keep your account secure with MetaMask Snaps"
   },
+  "addSnapAccountToggle": {
+    "message": "Enable \"Add snap account\""
+  },
+  "addSnapAccountsDescription": {
+    "message": "Turning on this feature will give you the option to add a Snap account right from your account list. If you install a Snap account, remember that it is a third-party service."
```
[the remark this answers is not in the document]
Comment (danroc) [reply app/_locales/en/messages.json:275]: It's from @coreyjanssen's [proposal](https://github.com/MetaMask/accounts-planning/issues/45#issuecomment-1707157453).

**Comment by danroc** [comment]
@HowardBraham, it's strange. I tried today to reproduce it but wasn't able to.

I added an E2E test, I'm hoping that if it happens again it will be caught by the CI.

File: app/_locales/en/messages.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (coreyjanssen) [reply app/_locales/en/messages.json:275]: We could remove "right" from the line, if that makes it more clear.

[the remark this answers is not in the document]
Comment (gantunesr) [reply app/_locales/en/messages.json:275]: It took me a while, but now I realized that  `account list` refers to the component. Maybe its just me, so no worries

**Comment by coreyjanssen** [comment]
A couple of capitalization issues on the screens. Can we do the following?

Change **Snap Accounts** to **Snap accounts**

Change **Enable snap accounts** to **Enable Snap accounts**

**Review by owencraston — APPROVED** [review]
Tested and it seems to be working as expected for me when I build flask. The code also looks good. Nice tests ✅ 

Screenshot 2023-09-21 at 11 34 45 AM

**Review by gantunesr — APPROVED** [review]

File: test/e2e/tests/settings-add-snap-account-toggle.spec.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        await driver.press('#password', driver.Key.ENTER);
+
+        // Make sure the "Add snap account" button is not visible.
+        await driver.clickElement('[data-testid="account-menu-icon"]');
+        await driver.assertElementNotPresent({
+          text: 'Add snap account',
+          tag: 'button',
+        });
+        await driver.clickElement('.mm-box button[aria-label="Close"]');
+
```
Comment (plasmacorral) [inline test/e2e/tests/settings-add-snap-account-toggle.spec.js:25]: For the ease of troubleshooting potential test failures going forward, will recommend that we consider separating this single block into two.

it ('add snap account disabled by default`...
and 
it ('add snap account after experimental setting enabled'....

File: test/e2e/tests/settings-add-snap-account-toggle.spec.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          '[data-testid="account-options-menu-button"]',
+        );
+        await driver.clickElement({ text: 'Settings', tag: 'div' });
+        await driver.clickElement({ text: 'Experimental', tag: 'div' });
+
+        // Switch "Enable Add snap account" to on.
+        const toggle = await driver.findElement(
+          '[data-testid="add-snap-account-toggle"]',
+        );
+        await driver.scrollToElement(toggle);
```
Comment (plasmacorral) [inline test/e2e/tests/settings-add-snap-account-toggle.spec.js:37]: We could probably turn this into a fixure and save test steps here.

File: test/e2e/tests/settings-add-snap-account-toggle.spec.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  it('switch "Enable Add snap account" to on', async function () {
+    await withFixtures(
+      {
+        fixtures: new FixtureBuilder().build(),
+        title: this.test.title,
+        failOnConsoleError: false,
+      },
+      async ({ driver }) => {
+        await driver.navigate();
+        await driver.fill('#password', 'correct horse battery staple');
```
Comment (plasmacorral) [inline test/e2e/tests/settings-add-snap-account-toggle.spec.js:15]: Can replace these two lines with `await unlockWallet(driver);` going forward

Comment (plasmacorral) [inline test/e2e/tests/settings-add-snap-account-toggle.spec.js:15]: This line and the next can be replaced by `await unlockWallet(driver);` going forward

**Comment by plasmacorral** [comment]
Manual tests primarily performed on Mac Ventura 13.5.2 in chrome 117 and firefox 117.  Can also confirm add snap account is disabled on first launch in Firefox 117 on Windows 11.

Made some minor suggestions/nits for consideration on the e2e test.  Will note that this works as described.  However, while this experimental setting is disabled there is nothing to stop a user from appending to the URL in expanded view to see account snap grid or to prevent accounts from being added to flask if the user navigates directly to a dapp that a user connects to that enables them to install the account snap directly and create new accounts.

**Pull request merged**